riscv-page-tables/src/page_table.rs

```diff
@@ -1207,6 +1283,137 @@ impl<T: PagingMode> GuestStagePageTable<T> {
         }))
     }
 
+    /// Verifies the entire virtual address range is invalidated and that each page is one size
+    /// smaller than the requested page size. Also checks the pages are physically contiguous and
+    /// suitably aligned to the requested page size. It's important that all the pages must be in
+    /// the same state, either "Blocked" and uniquely owned by the TVM, or "BlockedShared" and not
+    /// owned by the TVM. Last thing to check is that each page must have been invalidated with a
+    /// TLB version that's older than the current one.
+    pub fn promote_page<F>(
+        &self,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        requested_page_size: PageSize,
+        mut pred: F,
+    ) -> Result<(SupervisorPageAddr, PageSize, Page<InternalClean>)>
+    where
+        F: FnMut(SupervisorPageAddr, PageSize) -> bool,
+    {
+        let expected_page_size = requested_page_size
+            .size_down()
+            .ok_or(Error::InvalidPageSize)?;
+        let num_pages = PageSize::num_4k_pages(requested_page_size as u64);
+        let end = vaddr
+            .checked_add_pages(num_pages)
+            .ok_or(Error::AddressOverflow)?;
```

Comment:
need to check that `vaddr` is aligned. if it is, you're guaranteed not to overflow.